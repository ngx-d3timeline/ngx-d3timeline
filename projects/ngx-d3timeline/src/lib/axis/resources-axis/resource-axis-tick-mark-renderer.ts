import { BandScale } from '../../scale-types';
import { TickMarkRenderer, tickLabelSpacing } from '../tick-mark-renderer';
import { Orientation } from '../../orientation';

export function getResourceAxisTickMarkRenderer(
  scale: BandScale,
  orientation: Orientation
): TickMarkRenderer {
  const tickLineOffset = 0;

  function getBandMidPoint(tickValue: string): number {
    return scale(tickValue) + scale.bandwidth() / 2;
  }

  return {
    getTickLabel: (tickValue: string) => tickValue,
    getTickLabelSpacing: () => tickLineOffset + tickLabelSpacing,
    getTickValues: () => scale.domain(),
    mapTickValueToPositionInScale: (tickValue: string) =>
      getBandMidPoint(tickValue),
    tickLineOffset,
    orientation
  };
}
