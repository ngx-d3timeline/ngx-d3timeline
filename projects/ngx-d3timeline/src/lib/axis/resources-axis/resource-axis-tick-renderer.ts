import { BandScale } from '../../scale-types';
import { TickRenderer } from '../tick-renderer';

export class ResourceAxisTickRenderer implements TickRenderer<BandScale> {
  getTickValues(scale: BandScale): string[] {
    return scale.domain();
  }

  getLabel(_: BandScale, tick: any): string {
    return tick;
  }

  getTransform(scale: BandScale, tick: any): number {
    return this.getBandMidPoint(scale, tick);
  }

  private getBandMidPoint(scale: BandScale, tick: string): number {
    return scale(tick) + scale.bandwidth() / 2;
  }

  getTickLineOffset(): number {
    return 0;
  }
}
