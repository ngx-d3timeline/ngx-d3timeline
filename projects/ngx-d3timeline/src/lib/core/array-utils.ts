export function sum(...args: number[]) {
  return args.reduce((a, b) => a + b);
}
