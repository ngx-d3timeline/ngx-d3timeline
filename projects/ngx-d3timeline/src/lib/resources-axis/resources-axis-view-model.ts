import { ScaleBand, scaleBand } from 'd3-scale';
import { TimelineEvent } from '../timeline-event';
import { TimelineView } from '../view/timeline-view';
import { Orientation } from '../orientation';

export class ResourcesAxisViewModel {
  private readonly scaleBand: ScaleBand<string>;

  constructor(
    data: TimelineEvent[],
    timelineView: TimelineView,
    private readonly orientation: Orientation
  ) {
    this.scaleBand = this.configureScaleBand(data, timelineView);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.scaleBand.domain().map(tick => ({
      label: tick,
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.scaleBand.range()[1];
  }

  get bandwidth() {
    return this.scaleBand.bandwidth();
  }

  private configureScaleBand(
    data: TimelineEvent[],
    timelineView: TimelineView
  ) {
    return scaleBand()
      .domain([...new Set(data.map(d => d.series))])
      .range(this.getRange(timelineView.bounds));
  }

  private getRange(bounds: any): [number, number] {
    return this.orientation === Orientation.Vertical
      ? [bounds.left, bounds.right]
      : [bounds.top, bounds.bottom];
  }

  private tickTransform(tick: string) {
    return this.orientation === Orientation.Vertical
      ? `translate(${this.getBandMidPoint(tick)}, 0)`
      : `translate(0, ${this.getBandMidPoint(tick)})`;
  }

  private getBandMidPoint(tick: string) {
    return this.mapToPixel(tick) + this.scaleBand.bandwidth() / 2;
  }

  mapToPixel(domain: string): number {
    return this.scaleBand(domain);
  }
}
