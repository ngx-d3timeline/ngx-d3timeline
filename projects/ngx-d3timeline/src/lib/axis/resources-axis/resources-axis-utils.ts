import { TimelineView } from '../../view/timeline-view';
import { Orientation } from '../../orientation';
import { BandScale } from '../../scale-types';
import { ResourceAxisTickRenderer } from './resource-axis-tick-renderer';
import { createSelector } from '../../selector/memoized-selector';
import { selectView, selectResourceOrientation } from '../../store/state';
import { selectBandScale } from '../../store/timeline-selectors';
import { createAxis } from '../axis-utils';

function createResourceAxis(
  view: TimelineView,
  orientation: Orientation,
  bandScale: BandScale
) {
  return createAxis(
    view,
    orientation,
    bandScale,
    new ResourceAxisTickRenderer()
  );
}

export const selectResourceAxis = createSelector(
  [selectView, selectResourceOrientation, selectBandScale],
  createResourceAxis
);
