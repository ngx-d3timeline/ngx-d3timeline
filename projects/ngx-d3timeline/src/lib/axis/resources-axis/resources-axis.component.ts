import { Component, ChangeDetectionStrategy } from '@angular/core';
import { AxisService } from '../axis.service';

@Component({
  selector: '[ngx-d3timeline-resources-axis]',
  template: `
    <svg:g
      class="resources-axis-group"
      *ngIf="axisService.resourceAxis$ | async as axis"
    >
      <text
        *ngFor="let tick of axis.ticks"
        [attr.transform]="tick.transform"
        dy="-2"
      >
        {{ tick.label }}
      </text>
      <svg:g ngx-d3timeline-line [line]="axis.axisLine"></svg:g>
    </svg:g>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ResourcesAxisComponent {
  constructor(public axisService: AxisService) {}
}
