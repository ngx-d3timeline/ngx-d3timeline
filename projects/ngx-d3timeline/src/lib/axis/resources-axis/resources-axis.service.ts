import { Axis } from '../axis';
import { Orientation } from '../../orientation';
import { map } from 'rxjs/operators';
import { ScaleBand } from 'd3-scale';
import { Injectable } from '@angular/core';
import { Tick } from '../tick';
import { ScalesService } from '../../scales.service';
import { OptionsService } from '../../options.service';
import { AxisService } from '../axis.service';

@Injectable({ providedIn: 'root' })
export class ResourcesAxisService {
  vm$ = this.scalesService.scales$.pipe(
    map(scales =>
      this.createAxis(scales.scaleBand, scales.state.timelineOrientation)
    )
  );

  constructor(
    private scalesService: ScalesService,
    private optionsService: OptionsService,
    private axisService: AxisService
  ) {}

  private createAxis(
    scaleBand: ScaleBand<string>,
    timelineOrientation: Orientation
  ): Axis {
    const orientation = this.optionsService.flipOrientation(
      timelineOrientation
    );

    return {
      ticks: this.getTicks(scaleBand, orientation),
      axisLine: this.axisService.getAxisLine(scaleBand, orientation)
    };
  }

  private getTicks(scale: ScaleBand<string>, orientation: Orientation): Tick[] {
    return scale.domain().map(value => ({
      label: value,
      transform: this.optionsService.getTranslation(
        this.getBandMidPoint(scale, value),
        orientation
      )
    }));
  }

  private getBandMidPoint(scale: ScaleBand<string>, tick: string) {
    return scale(tick) + scale.bandwidth() / 2;
  }
}
