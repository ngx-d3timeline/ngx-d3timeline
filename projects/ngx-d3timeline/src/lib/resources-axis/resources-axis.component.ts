import { Component, Input } from '@angular/core';
import { ResourcesAxisViewModel } from './resources-axis-view-model';
import { AxisViewService } from '../scale/axis-view.service';

@Component({
  selector: '[ngx-d3timeline-resources-axis]',
  template: `
    <svg:g
      class="resources-axis-group"
      *ngIf="axisViewService.resourceAxisVm$ | async as vm"
    >
      <text
        *ngFor="let tickInfo of vm.tickInfos"
        [attr.transform]="tickInfo.transform"
        dy="-2"
      >
        {{ tickInfo.label }}
      </text>
      <svg:line
        class="resources-axis-line"
        [attr.x2]="vm.axisLine.x2"
        [attr.y2]="vm.axisLine.y2"
      ></svg:line>
    </svg:g>
  `,
  styles: [
    `
      .resources-axis-group {
        text-anchor: middle;
      }
      .resources-axis-line {
        stroke: #000;
      }
    `
  ]
})
export class ResourcesAxisComponent {
  constructor(public axisViewService: AxisViewService) {}
}
