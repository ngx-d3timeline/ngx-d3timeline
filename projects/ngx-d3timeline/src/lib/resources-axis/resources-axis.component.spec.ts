import { TestBed, ComponentFixture } from '@angular/core/testing';
import { ResourcesAxisComponent } from './resources-axis.component';
import { ResourcesAxisService } from './resources-axis.service';
import { of } from 'rxjs';

describe('ResourcesAxisComponent', () => {
  let fixture: ComponentFixture<ResourcesAxisComponent>;
  let axisService: ResourcesAxisService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ResourcesAxisComponent],
      providers: [
        { provide: ResourcesAxisService, useValue: { vm$: jest.fn() } }
      ]
    });

    fixture = TestBed.createComponent(ResourcesAxisComponent);
    axisService = TestBed.inject(ResourcesAxisService);
  });

  it('should not render if view model is null', () => {
    axisService.vm$ = of(null);
    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render correctly', () => {
    axisService.vm$ = of({
      tickInfos: [
        {
          label: 'tick 1',
          transform: 'translate(0, 10)'
        },
        {
          label: 'tick 2',
          transform: 'translate(0, 20)'
        }
      ],
      axisLine: { x1: 0, x2: 10, y1: 1, y2: 0 }
    });

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
