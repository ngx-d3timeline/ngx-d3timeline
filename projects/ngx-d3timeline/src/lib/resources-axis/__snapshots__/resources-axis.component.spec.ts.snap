// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourcesAxisComponent should not render if view model is null 1`] = `
<div
  id="root0"
  ngx-d3timeline-resources-axis=""
>
  
</div>
`;

exports[`ResourcesAxisComponent should render correctly 1`] = `
<div
  id="root1"
  ngx-d3timeline-resources-axis=""
>
  
  <g
    class="resources-axis-group"
  >
    
    <text
      dy="-2"
      transform="translate(116.66666666666667, 0)"
    >
       Series 1 
    </text>
    <text
      dy="-2"
      transform="translate(350, 0)"
    >
       Series 2 
    </text>
    <text
      dy="-2"
      transform="translate(583.3333333333334, 0)"
    >
       Series 3 
    </text>
    <line
      class="resources-axis-line"
      x2="700"
    />
  </g>
</div>
`;
